tensorboard/plugins/graph/tf_graph_common/hierarchy.ts

```diff
@@ -885,22 +890,70 @@ function detectSeries(clusters: {[clusterId: string]: string[]},
       let namepath = name.split('/');
       let leaf = namepath[namepath.length - 1];
       let parent = namepath.slice(0, namepath.length - 1).join('/');
-      let matches = leaf.match(/^(\D*)_(\d+)$/);
 
+      const numRegex = /(\d+)/g;
+      let matches = [];
+      let matchResult;
       let prefix;
       let id;
-      let suffix = '';
-      if (matches) {         // if found '<number>' in the name, assign id.
-        prefix = matches[1]; // the front non-numeric characters
-        id = matches[2]; // the digits
-      } else {  // for node without '_<number>', make them zero-th items.
+      let suffix;
+      let seriesName;
+      let matched = 0;
+      // Scan over the entire leaf name and match any possible numbers,
+      // and put the results into corresponding dictionaries.
+      while (matchResult = numRegex.exec(leaf)) {
```

Comment:
What you done any perf measurement? Curious what the perf impact is of this change.

Reply:
I did some simple benchmarks on a couple of graphs and noticed that this new logic for series collapsing may change some existing behavior. It may also have a performance impact on graphs with many ops within a single hierarchy, although I wasn't able to measure a significant performance difference with the graphs I was benchmarking with.

I therefore guarded this feature behind a `useGeneralizedSeriesPatterns` hierarchy param. I also enabled developers to override hierarchy parameters. See latest commit. This seems like a safe way to introduce this logic.